### PR TITLE
SearchBuilder should not allow $everything operation to follow links to Group or List resources

### PR DESCRIPTION
**Pull request by volodymyr-korzh**
Modified SearchBuilder so that $everything doesn't follow links to Group or List resources, so they will never be included in the $everything response.

It is now possible to configure permission rules for operation with Access All rights.

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 1]
```diff
@@ -1381,6 +1381,8 @@ public Set<JpaPid> loadIncludes(SearchBuilderLoadIncludesParameters<JpaPid> theP
 							&& myParams.getEverythingMode() == SearchParameterMap.EverythingModeEnum.PATIENT_INSTANCE) {
 						sqlBuilder.append(" AND r.myTargetResourceType != 'Patient'");
 						sqlBuilder.append(" AND r.mySourceResourceType != 'Provenance'");
+						sqlBuilder.append(" AND r.mySourceResourceType != 'List'");
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:1384]: I notice you added this to the `UNDESIRED_RESOURCE_LINKAGES_FOR_PATIENT_INSTANCE` constant in the corresponding Smile MR - would it make sense to extract this list of resources into a HAPI FHIR constant that can be referred to in both places?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderOperationNamedAndScoped.java [hunk 2]
```diff
@@ -26,6 +26,11 @@ public interface IAuthRuleBuilderOperationNamedAndScoped {
 	 */
 	IAuthRuleBuilderRuleOpClassifierFinished andAllowAllResponses();
 
+	/**
+	 * Responses for this operation will not be checked and access to all resources is allowed
+	 */
```
Comment (jamesagnew) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderOperationNamedAndScoped.java:31]: Suggesting a small beef-up of this comment just to make the intent clear.

```suggestion
	 * Responses for this operation will not be checked and access to all resources is allowed. This
	 * is intended for operations which are known to fetch a graph of resources that is known to be
	 * safe, such as `$everything` which may access and fetch resources outside the patient's compartment
	 * but enforces safety in what it fetches via strict SQL queries.
	 */
```

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4EverythingTest.java [hunk 3]
```diff
@@ -15,26 +15,7 @@
...
-import org.hl7.fhir.r4.model.Observation;
-import org.hl7.fhir.r4.model.Organization;
-import org.hl7.fhir.r4.model.Parameters;
-import org.hl7.fhir.r4.model.Patient;
-import org.hl7.fhir.r4.model.Practitioner;
-import org.hl7.fhir.r4.model.Provenance;
-import org.hl7.fhir.r4.model.Reference;
-import org.hl7.fhir.r4.model.StringType;
-import org.hl7.fhir.r4.model.UnsignedIntType;
+import org.hl7.fhir.r4.model.*;
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4EverythingTest.java:18]: Nitpick: Star import

**Review by jamesagnew — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 1 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: Extracted this list of resource types to JpaConstants - now this set is used in both HAPI SearchBuilder and in Smile changes

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4EverythingTest.java (hunk 3 above)
Comment (volodymyr-korzh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4EverythingTest.java]: Fixed

**Pull request merged**